File: PSN/Network/particle_swarm.cpp
#include "../psn.h"

void ParticleSwarm::path_reset()
{
  std::vector<NetworkConfiguration>::iterator net_it;
  for (net_it = paths.begin(); net_it < paths.end(); ++net_it) {
    *net_it = c_net;
  }
}

void ParticleSwarm::path_branch()
{
  std::vector<NetworkConfiguration>::iterator net_it;
  for (net_it = paths.begin(); net_it < paths.end(); ++net_it) {
    net_permute(*net_it);
    Network* net = new Network(*net_it);
    Network* optnet = new Network(c_net);

    net->run(32);
    optnet->run(32);

    if (net->fitness() >= optnet->fitness()) {
      c_net = *net_it;
    }

    delete net;
    delete optnet;
  }
}

void ParticleSwarm::init()
{
  n_paths = 20;
  for (unsigned int i = 0; i < n_paths; ++i) {
    paths.push_back(NetworkConfiguration());
  }
  net_permute(c_net);
}

ParticleSwarm::ParticleSwarm()
{
  init();
}

File: psnpy/Network/particle_swarm.py
import numpy as np
import network

class ParticleSwarm:
  def __init__(self):
    self.n_paths = 20
    self.paths = []
    self.c_net = network.Network()

  def path_reset(self):
    self.paths = []
    self.paths.append(self.c_net)

  def path_branch(self):
    for path in self.paths:
      path.mutate()
      path.run(2)

    for path in self.paths:
      if path.fitness() >= self.c_net.fitness():
        self.c_net = path

    self.paths = []

